share/CMakeJS.cmake

```
@@ -0,0 +1,642 @@
+    cmakejs_acquire_napi_c_files()
+    message(STATUS "NODE_API_HEADERS_DIR: ${NODE_API_HEADERS_DIR}")
+    if(NOT DEFINED NODE_API_INC_FILES)
+        file(GLOB NODE_API_INC_FILES "${NODE_API_HEADERS_DIR}/*.h")
+        set(NODE_API_INC_FILES "${NODE_API_INC_FILES}" CACHE FILEPATH "Node API Header files." FORCE)
+        source_group("Node Addon API (C)" FILES "${NODE_API_INC_FILES}")
+    endif()
+endif()
+
+#[=============================================================================[
+Get NodeJS C++ Addon development files.
+
+Provides
+::
+
+  NODE_ADDON_API_DIR, where to find napi.h, etc.
+  NODE_ADDON_API_INC_FILES, the headers required to use Node Addon API.
+
+]=============================================================================]#
+function(cmakejs_acquire_napi_cpp_files)
+    execute_process(
+      COMMAND "${NODE_EXECUTABLE}" -p "require('node-addon-api').include"
+      WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+      OUTPUT_VARIABLE NODE_ADDON_API_DIR
+      # COMMAND_ERROR_IS_FATAL ANY
+    )
+    string(REGEX REPLACE "[\r\n\"]" "" NODE_ADDON_API_DIR "${NODE_ADDON_API_DIR}")
+    set(NODE_ADDON_API_DIR "${NODE_ADDON_API_DIR}" CACHE PATH "Node Addon API Headers directory." FORCE)
+
+    unset(NODE_ADDON_API_INC_FILES)
+    file(GLOB NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_DIR}/*.h")
+    set(NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_INC_FILES}" CACHE FILEPATH "Node Addon API Header files." FORCE)
+    source_group("Node Addon API (C++)" FILES "${NODE_ADDON_API_INC_FILES}")
+
+    if(VERBOSE)
+        message(STATUS "NODE_ADDON_API_DIR: ${NODE_ADDON_API_DIR}")
+    endif()
+endfunction()
+
+# Acquire if needed...
+if(NOT DEFINED NODE_ADDON_API_DIR)
+    cmakejs_acquire_napi_cpp_files()
+    message(STATUS "NODE_ADDON_API_DIR: ${NODE_ADDON_API_DIR}")
+    if(NOT DEFINED NODE_ADDON_API_INC_FILES)
+	file(GLOB NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_DIR}/*.h")
+	set(NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_INC_FILES}" CACHE FILEPATH "Node Addon API Header files." FORCE)
+        source_group("Node Addon API (C++)" FILES "${NODE_ADDON_API_INC_FILES}")
+    endif()
+endif()
+
+#[=============================================================================[
+Silently create an interface library (no output) with all Addon API dependencies
+resolved, for Addon targets to link with.
+
+(This should contain most of cmake-js globally-required configuration)
+
+Targets:
+
+cmake-js::node-dev
+cmake-js::node-api
+cmake-js::node-addon-api
+cmake-js::cmake-js
+
+]=============================================================================]#
+
+# NodeJS system installation headers
+# cmake-js::node-dev
+add_library                 (node-dev INTERFACE)
+add_library                 (cmake-js::node-dev ALIAS node-dev)
+target_include_directories  (node-dev INTERFACE "${CMAKE_JS_INC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_SRC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_INC_FILES}")
```

```suggestion
if (MSVC)
    target_sources              (node-dev INTERFACE "${CMAKE_CURRENT_SOURCE_DIR}/node_modules/cmake-js/lib/cpp/win_delay_load_hook.cc")
endif()
```

If we ignore non node-api projects for now, then `CMAKE_JS_INC` is just giving the same includes as `cmake-js::node-api` and `cmake-js::node-addon-api`.

The only source file cmake-js injects is this one for windows. I don't like how manual the path is here, feel free to tidy it up.